### DISCUSSION TITLE
Problems with type inference

### DISCUSSION THREAD

**Discussion by GigaFirmDev** [tokio-rs/axum#2179] [Q&A] [answered]
Lately, I was dealing with this problem where rust cannot infer the type generics of the route. I even had to resort to axum examples to check if it was a mistake but the result stayed the same:

```rs
let app = Router::new()
        // `GET /` goes to `root`
        .route("/", get(root))
        // `POST /users` goes to `create_user`
        .route("/users", post(create_user));
```
 This is the portion of the code that is causing me trouble. This is what rust said about it:

```
error[E0283]: type annotations needed
   --> src/main.rs:14:15
    |
14  |     let app = Router::new()
    |               ^^^^^^^^^^^ cannot infer type of the type parameter `B` declared on the struct `Router`
    |
    = note: cannot satisfy `_: HttpBody`
    = help: the following types implement trait `HttpBody`:
              &mut T
              Body
              Box<T>
              Pin<P>
              Request<B>
              Response<B>
              StreamBody<S>
              axum::body::Empty<D>
            and 7 others
note: required by a bound in `Router::<S, B>::new`
   --> PATH
    |
105 |     B: HttpBody + Send + 'static,
    |        ^^^^^^^^ required by this bound in `Router::<S, B>::new`
...
112 |     pub fn new() -> Self {
    |            --- required by a bound in this associated function
help: consider specifying the generic arguments
    |
14  |     let app = Router::<S, B>::new()
    |                     ++++++++

For more information about this error, try `rustc --explain E0283`.
error: could not compile `backend` (bin "backend") due to previous error
```

**Comment by davidpdrsn** [comment, marked as answer]
If you actually run the router with

```rust
axum::Server::bind(&addr)
    .serve(app.into_make_service())
    .await
    .unwrap();
```

the compile is able to infer the types.

**Reply by GigaFirmDev** [reply]
Oh thanks. I knew it was a mistake of my part. Again thanks

**Reply by jplatte** [reply]
I'm not sure which part of the expression David posted makes the compiler infer the body type, but worst case you can always add a type annotation which will apply the defaults of `S = ()` and `B = Body`:

```rust
let app: Router = Router::new()
    // `GET /` goes to `root`
    .route("/", get(root))
    // `POST /users` goes to `create_user`
    .route("/users", post(create_user));
```

**Reply by davidpdrsn** [reply]
@jplatte The trait bounds on `serve` requires the request body type to be `Body`:

![Untitled](https://github.com/tokio-rs/axum/assets/718941/98e53b82-77da-4510-b7db-1f469a9d3868)

Though its not exactly obvious 😅

**Reply by jplatte** [reply]
I just saw that there's still a `B` parameter and assumed the method was generic. Both the use of `Body` and `B` being inside generic parameter lists of private types doesn't make this one easier to figure out 😬

**Reply by davidpdrsn** [reply]
I believe `B` is the response body.

**Reply by davidpdrsn** [reply]
> I just saw that there's still a B parameter and assumed the method was generic. Both the use of Body and B being inside generic parameter lists of private types doesn't make this one easier to figure out 😬

hehe yes

**Reply by jplatte** [reply]
So hyper determines the fixed request body, and axum determines the fixed response body. Really looking forward to hyper 1.0 / axum 0.7 to get most people off of this overly-generic function 😄
